### PR TITLE
89 Grid to DiGraph Conversion and Graph Simplification.

### LINKED ISSUE #19: Add a standard graph representation/transformation for rail environments
Many solution approaches (both RL and planning) did operate on a graph representation of flatland.
We have a least two versions of graphs somewhere, one of them based on the netcetera graph.

We should have a "standard" graph transformation function for flatland. Maybe placed in utils.
(For example, a modified version of the netcetera graph).

### PR DESCRIPTION
**Pull request by chenkins**
## Changes
* grid to digraph conversion
* data structures for simplified graph (decision point graph/overlay graph)

## Related issues
First steps for #89 
Resolves #19
Documentation see also flatland-association/flatland-book#6, in particular https://flatland-association.github.io/flatland-book/environment/graph_demo.html
## Checklist

- [x] Tests are included for relevant behavior changes.
- [x] Documentation is added in the `docs` folder for relevant behavior changes. If you made important user-facing
  changes, describe them under the `[Unreleased]` tag in `CHANGELOG.md`.
- [x] New package dependencies are declared in the `pyproject.toml` file.
  Requirement files have been updated by running `tox -e requirements`.
- [x] Code works with all supported Python versions (3.10, 3.11 and 3.12). Checks run with all three version and are
  required to run successfully.
- [x] Code is formatted according to PEP 8 (an IDE like PyCharm can do this for you).
- [x] Technical guidelines listed in `CONTRIBUTING.md` are followed.

### CODE REVIEW

File: flatland/core/graph/grid_to_graph.py [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+The implementation uses networkX, so there are also many graph functions available.
+
+References:
+- Egli, Adrian. FlatlandGraphBuilder. https://github.com/aiAdrian/flatland_railway_extension/blob/e2b15bdd851ad32fb26c1a53f04621a3ca38fc00/flatland_railway_extension/FlatlandGraphBuilder.py
+- Nygren, E., Eichenberger, Ch., Frejinger, E. Scope Restriction for Scalable Real-Time Railway Rescheduling: An Exploratory Study. https://arxiv.org/abs/2305.03574
+TODO split file: env creation, graph derivation, simplification, rendering, notebook
+TODO action into cell behaviour - add to documentation
+TODO docs with illustration of the mapping and "pins".
+TODO illustration simplification and edge cases (pun intended): non-facing switches and diamond-crossings, and multi-edges.
+TODO tests
```
Comment (chenkins) [inline flatland/core/graph/grid_to_graph.py:24]: TODOs

File: flatland/core/graph/grid_to_graph.py [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+from flatland.envs.observations import TreeObsForRailEnv
+from flatland.envs.predictions import ShortestPathPredictorForRailEnv
+from flatland.envs.rail_env import RailEnv
+from flatland.envs.rail_generators import sparse_rail_generator
+from utils.rendertools import RenderTool
+
+GridNode = Tuple[int, int, int]
+
+
+# TODO move method
```
Comment (chenkins) [inline flatland/core/graph/grid_to_graph.py:47]: TODO

File: flatland/core/graph/grid_to_graph.py [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+        ),
+        malfunction_generator=ParamMalfunctionGen(MalfunctionParameters(
+            min_duration=malfunction_duration_min, max_duration=malfunction_duration_max, malfunction_rate=1.0 / malfunction_interval)),
+        line_generator=sparse_line_generator(speed_ratio_map=speed_ratios, seed=seed),
+        number_of_agents=n_agents,
+        obs_builder_object=obs_builder_object,
+        record_steps=True,
+        random_seed=seed
+    )
+    # TODO check env reset is deterministic?
```
Comment (chenkins) [inline flatland/core/graph/grid_to_graph.py:85]: TODO

File: flatland/core/graph/grid_to_graph.py [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+    def fromRailEnv(env: RailEnv) -> "GraphTransitionMap":
+        return GraphTransitionMap(GraphTransitionMap.grid_to_digraph(env.rail))
+
+
+@attrs
+class DecisionPointGraphEdgeData:
+    path = attrib(type=List[GridNode])
+
+
+# TODO naming: decision point graph, simplified graph, collapsed graph, ....?
```
Comment (chenkins) [inline flatland/core/graph/grid_to_graph.py:127]: TODO

File: flatland/core/graph/grid_to_graph.py [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+
+        # add edge for
+        for dp in decision_nodes:
+            for n in micro.successors(dp):
+                branch = DecisionPointGraph._explore_branch(micro, dp, n)
+                g.add_edge(branch[0], branch[-1], d=DecisionPointGraphEdgeData(path=branch))
+        return DecisionPointGraph(g)
+
+
+# TODO separation of concerns: graph building and rendering...
```
Comment (chenkins) [inline flatland/core/graph/grid_to_graph.py:160]: TODO

File: flatland/core/graph/grid_to_graph.py [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+    ax.set_yticks(np.arange(0, env.height, 1))
+    # TODO image does not fill extent entirely - why?
+    ax.imshow(np.fliplr(np.rot90(np.rot90(img))), extent=[-0.5, env.width - 0.5, -0.5, env.height - 0.5])
+    ax.set_xticks(np.arange(-0.5, env.width + 0.5, 1), minor=True)
+    ax.set_yticks(np.arange(-0.5, env.height + 0.5, 1), minor=True)
+    ax.tick_params(left=True, bottom=True, labelleft=True, labelbottom=True)
+    ax.grid(which="minor")
+
+
+# TODO move to notebook
```
Comment (chenkins) [inline flatland/core/graph/grid_to_graph.py:189]: TODO

**Comment by aiAdrian** [comment]
> grafik grafik

Just wonder - what does the left flatland-graph show? Is there the "big loop e.g. top left to middle center" not missing? Micro looks very good. But the left one i don't understand from visualisation.

File: flatland/core/graph/graph_simplification.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+    len: int
+        Number of collapsed cells.
+
+    """
+    path = attrib(type=List[GridNode])
+    len = attrib(type=int)
+
+
+# TODO naming: decision point graph, simplified graph, collapsed graph, ....?
+class DecisionPointGraph:
```
Comment (chenkins) [inline flatland/core/graph/graph_simplification.py:29]: @aiAdrian is this an appropriate naming? decision point graph, simplified graph, collapsed graph, ....?

File: tests/test_graph_simplification.py [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+from flatland.core.graph.grid_to_graph import GraphTransitionMap
+from flatland.utils.simple_rail import make_simple_rail, make_disconnected_simple_rail, make_simple_rail2, make_simple_rail_with_alternatives, make_oval_rail
+
+
+@pytest.mark.parametrize("env_creator,expected_num_decision_points", [
+    (make_simple_rail, 2),
+    (make_disconnected_simple_rail, 2),
+    (make_simple_rail2, 2),
+    (make_simple_rail_with_alternatives, 2),
+    (make_oval_rail, 0),  # TODO what do we want here?
```
Comment (chenkins) [inline tests/test_graph_simplification.py:13]: @aiAdrian edge case - what's your take on this?

File: flatland/core/graph/grid_to_graph.py [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+
+Two (dual, equivalent) approaches are possible:
+- agents are positioned on the nodes
+- agents are positioned on the edges.
+The second approach makes it easier to visualize agents moving forward on edges. Hence, we choose the second approach.
+
+Our directed graph consists of nodes and edges:
+* A node in the graph is defined by position and direction. The position corresponds to the position of the underlying cell in the original flatland topology, and the direction corresponds to the direction in which an agent reaches the cell. Thus, the node is defined by (r, c, d), where c (column) is the index of the horizontal cell grid position, r (row) is the index of the vertical cell grid position, and d (direction) is the direction of cell entry. In the Flatland (2d grid), not every of the eight neighbors cell can be reached from every direction. Therefore, the entry direction information is key.
```
Comment (SergeCroise) [inline flatland/core/graph/grid_to_graph.py:12]:
```suggestion
* A node in the graph is defined by position and direction. The position corresponds to the position of the underlying cell in the original flatland topology, and the direction corresponds to the direction in which an agent reaches the cell. Thus, the node is defined by (r, c, d), where c (column) is the index of the horizontal cell grid position, r (row) is the index of the vertical cell grid position, and d (direction) is the direction of cell entry. In the Flatland (2d grid), not every of the eight neighbors cells can be reached from every direction. Therefore, the entry direction information is key.
```

File: flatland/core/graph/grid_to_graph.py [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+* An edge is defined by "from-node" u and "to-node" v such that for the edge e = (u, v).  Edges reflect feasible transition from node u to node v exist.
```
Comment (SergeCroise) [inline flatland/core/graph/grid_to_graph.py:13]:
```suggestion
* An edge is defined by "from-node" u and "to-node" v such that for the edge e = (u, v).  Edges reflect feasible transition from node u to node v.
```

**Comment by hagrid67** [comment]
I haven't read all the documentation here but I think we need to be explicit that, in this formulation, at any given time, an agent "occupies" an edge (not a node).  The step() call results in the agent transitioning from one edge to another edge.  A node with two out-edges is therefore a decision node (vs a single out-edge); an agent on the "predecessor" edge needs to decide which "successor" edge to take as it crosses the decision node.  (The two successor edges are both "in" the same cell)

It's then not so obvious which row and column a *node* is on, because it is on the boundary between two cells.  A cell has a unique (row, column). An edge is "in" a single cell (row, column).  Either the row or column of a node seems ambiguous (a node is on either a horizontal or vertical boundary so only one is ambiguous.  We have no diagonal transition between cells.)

Since we define (r,c,d) with d is direction of *entry*, this resolves the ambiguity. However it means that the twin points on the same boundary are split across the adjacent cells, eg:
- The two nodes on East/West rails between cells (3,4) and (3,5).  The "East" node is (3,5,East) and the West node is (3,4,West), because we *enter* 3,5 going East and we *enter* 3,4 going West.
- if (3,4) is a switch, say going North to (2,4) and West to (3,3), then we have edges
-- (3,4,West) -> (3,3,West)
-- (3,4,West) -> (2,4,North)

Ideally this should be represented in the diagram, ie the nodes should be slightly offset from the cell boundaries.

**Comment by chenkins** [comment]
@hagrid67 The implementation of this pr only reflects the topology, there are no agents (yet) "living" on this graph, no `step()` working with this graph yet. This is future/ongoing work not part of this pr. But you're right that agents will live on the edges.

There are only "entry nodes", no "exit nodes", there are no "twin points". The grey dots in Figure 5 do not actually exist, but are proxy for the entry node in the neighboring cell, nodes just separating two edges and reflecting the possible transitions.

The implementation is pretty straightforward: for all transitions in the transition map, add an edge to the graph:
https://github.com/flatland-association/flatland-rl/blob/e33821b471527e914055adf815a90fff342fe824/flatland/core/graph/grid_to_graph.py#L83-L92

See also https://github.com/flatland-association/flatland-book/blob/68e6aafe6a5297398b57e419c7a9b025ac0ff918/environment/graph_demo.ipynb

File: notebooks/graph_demo.ipynb [hunk 11]
```diff
@@ -0,0 +1,392 @@
...
+    "\n",
+    "Two (dual, equivalent) approaches are possible:\n",
+    "- agents are positioned on the nodes\n",
+    "- agents are positioned on the edges.\n",
+    "The second approach makes it easier to visualize agents moving forward on edges. Hence, we choose the second approach.\n",
+    "\n",
+    "Our directed graph consists of nodes and edges:\n",
+    "* A node in the graph is defined by position and direction. The position corresponds to the position of the underlying cell in the original flatland topology, and the direction corresponds to the direction in which an agent reaches the cell. Thus, the node is defined by (r, c, d), where c (column) is the index of the horizontal cell grid position, r (row) is the index of the vertical cell grid position, and d (direction) is the direction of cell entry. In the Flatland (2d grid), not every of the eight neighbor cells can be reached from every direction. Therefore, the entry direction information is key.\n",
+    "* An edge is defined by \"from-node\" u and \"to-node\" v such that for the edge e = (u, v).  Edges reflect feasible transition from node u to node v.\n",
```
Comment (chenkins) [inline notebooks/graph_demo.ipynb:71]:
```suggestion
    "* An edge is defined by \"from-node\" u and \"to-node\" v such that for the edge e = (u, v).  Edges reflect feasible transition from node u to node v. We can think of the suggestive notation $[u,v)$ in terms of resource occupation of the underlying cell, as the \"from-node\" refers to the underlying cell entered, and the \"to-node\" refers to the neighbor cell entered when the edge is left.\n",

```

**Review by aiAdrian — APPROVED** [review]

File: flatland/core/graph/graph_simplification.py [hunk 12]
```diff
@@ -74,9 +77,24 @@ def fromGraphTransitionMap(gtm: GraphTransitionMap) -> "DecisionPointGraph":
...
+
+        # special cases closed loops
+        open = set(micro.edges) - closed
+        while not len(open) == 0:
+            u, v = next(iter(open))
+            print((u, v))
+            branch = DecisionPointGraph._explore_branch(micro, u, v)
+            g.add_edge(branch[0], branch[-1], d=DecisionPointGraphEdgeData(path=branch, len=len(branch)))
+            for u_, v_, in zip(branch, branch[1:]):
+                print(f" rmove ({(u_, v_)}")
```
Comment (aiAdrian) [inline flatland/core/graph/graph_simplification.py:97]: Why - printf - is this for debug purpose only? This would take quite a lot of performance. If this could be removed.

**Review by aiAdrian — DISMISSED** [review]

**Review by aiAdrian — APPROVED** [review]
